// arui-feather/tab-item/tab-item.js
"use strict";
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var bem_react_classname_1 = require("bem-react-classname");
var cn_1 = require("../cn");
var link_1 = require("../link/link");
/**
 * Компонент таба. Как правило, используется совместно с `Tabs`.
 */
var TabItem = /** @class */ (function (_super) {
    __extends(TabItem, _super);
    function TabItem() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.cn = bem_react_classname_1.createCn('tab-item');
        return _this;
    }
    TabItem.defaultProps = {
        size: 'l',
        view: 'default',
        disabled: false,
        checked: false,
        pseudo: false,
        tabIndex: 0,
    };
    return TabItem;
}(link_1.Link));
exports.TabItem = TabItem;
exports.default = cn_1.withTheme(TabItem);
