// arui-feather/form/form.js
"use strict";
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var bem_react_classname_1 = require("bem-react-classname");
var cn_1 = require("../cn");
/**
 * Компонент формы.
 */
var Form = /** @class */ (function (_super) {
    __extends(Form, _super);
    function Form() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.cn = bem_react_classname_1.createCn('form');
        _this.handleSubmit = function (event) {
            event.preventDefault();
            if (_this.props.onSubmit) {
                _this.props.onSubmit(event);
            }
        };
        return _this;
    }
    Form.prototype.render = function () {
        return (react_1.default.createElement("form", { action: this.props.action, encType: this.props.enctype, method: this.props.method, noValidate: this.props.noValidate, autoComplete: this.props.autocomplete === false ? 'off' : 'on', className: this.cn({
                size: this.props.size,
                view: this.props.view,
            }), id: this.props.id, name: this.props.name, onSubmit: this.handleSubmit, "data-test-id": this.props['data-test-id'] },
            this.props.children,
            this.props.footer
                && (react_1.default.createElement("div", { className: this.cn('footer') }, this.props.footer))));
    };
    Form.defaultProps = {
        action: '/',
        enctype: 'application/x-www-form-urlencoded',
        method: 'post',
        size: 'm',
        noValidate: false,
        autocomplete: true,
    };
    return Form;
}(react_1.default.PureComponent));
exports.Form = Form;
exports.default = cn_1.withTheme(Form);
