// arui-feather/card-input/card-input.js
"use strict";
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var bem_react_classname_1 = require("bem-react-classname");
var cn_1 = require("../cn");
var input_1 = require("../input/input");
/**
 * Поле ввода номера карты с маской.
 */
var CardInput = /** @class */ (function (_super) {
    __extends(CardInput, _super);
    function CardInput() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.cn = bem_react_classname_1.createCn('card-input');
        return _this;
    }
    CardInput.prototype.render = function () {
        var _this = this;
        return (react_1.default.createElement(input_1.Input, __assign({}, this.props, { type: "tel", ref: function (root) {
                _this.root = root;
            }, mask: "1111 1111 1111 1111 11", formNoValidate: true, className: this.cn() })));
    };
    /**
     * Устанавливает фокус на поле ввода.
     */
    CardInput.prototype.focus = function () {
        this.root.focus();
    };
    /**
     * Убирает фокус с поля ввода.
     */
    CardInput.prototype.blur = function () {
        this.root.blur();
    };
    /**
     * Скроллит страницу до поля ввода.
     */
    CardInput.prototype.scrollTo = function () {
        this.root.scrollTo();
    };
    CardInput.defaultProps = {
        placeholder: '0000 0000 0000 0000 00',
    };
    return CardInput;
}(react_1.default.PureComponent));
exports.CardInput = CardInput;
exports.default = cn_1.withTheme(CardInput);
