// arui-feather/mq/decorator.js
"use strict";
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
var react_1 = __importDefault(require("react"));
var match_media_1 = require("../lib/match-media");
function mqDecorator(query, propName) {
    if (propName === void 0) { propName = 'mqMatch'; }
    return function (Component) {
        return /** @class */ (function (_super) {
            __extends(class_1, _super);
            function class_1() {
                var _this = _super !== null && _super.apply(this, arguments) || this;
                _this.mql = null;
                _this.state = {
                    matched: false,
                };
                _this.handleMatchChange = function (mql) {
                    _this.setState({
                        matched: mql.matches,
                    });
                };
                return _this;
            }
            class_1.prototype.componentDidMount = function () {
                this.mql = match_media_1.getMatchMedia(query);
                this.handleMatchChange(this.mql);
                // TODO: depricated => change to addEventListener
                this.mql.addListener(this.handleMatchChange);
            };
            class_1.prototype.componentWillUnmount = function () {
                match_media_1.releaseMatchMedia(query);
                // TODO: depricated => change to removeEventListener
                this.mql.removeListener(this.handleMatchChange);
                this.mql = null;
            };
            class_1.prototype.render = function () {
                var _a;
                return react_1.default.createElement(Component, __assign(__assign({}, this.props), (_a = {}, _a[propName] = this.state.matched, _a)));
            };
            return class_1;
        }(react_1.default.Component));
    };
}
exports.default = mqDecorator;
